File: src/main/java/com/github/dylmeadows/eontimer/core/TimerState.java
package com.github.dylmeadows.eontimer.core;

import com.github.dylmeadows.eontimer.model.Stage;
import javafx.beans.property.ReadOnlyBooleanProperty;
import javafx.beans.property.ReadOnlyBooleanWrapper;
import javafx.beans.property.ReadOnlyIntegerProperty;
import javafx.beans.property.ReadOnlyIntegerWrapper;
import javafx.beans.property.ReadOnlyLongProperty;
import javafx.beans.property.ReadOnlyLongWrapper;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.property.ReadOnlyObjectWrapper;
import org.springframework.stereotype.Component;

import java.util.Objects;

@Component
@SuppressWarnings({"unused", "WeakerAccess"})
public class TimerState {

    private final ReadOnlyLongWrapper remaining = new ReadOnlyLongWrapper();
    private final ReadOnlyObjectWrapper<Stage> currentStage = new ReadOnlyObjectWrapper<>();
    private final ReadOnlyObjectWrapper<Stage> nextStage = new ReadOnlyObjectWrapper<>();
    private final ReadOnlyIntegerWrapper minutesBeforeTarget = new ReadOnlyIntegerWrapper();
    private final ReadOnlyBooleanWrapper running = new ReadOnlyBooleanWrapper();

    public long getRemaining() {
        return remaining.get();
    }

    public ReadOnlyLongProperty remainingProperty() {
        return remaining.getReadOnlyProperty();
    }

    void setRemaining(long remaining) {
        this.remaining.set(remaining);
    }

    public Stage getCurrentStage() {
        return currentStage.get();
    }

    public ReadOnlyObjectProperty<Stage> currentStageProperty() {
        return currentStage.getReadOnlyProperty();
    }

    void setCurrentStage(Stage currentStage) {
        Objects.requireNonNull(currentStage);
        this.currentStage.set(currentStage);
    }

    public Stage getNextStage() {
        return nextStage.get();
    }

    public ReadOnlyObjectProperty<Stage> nextStageProperty() {
        return nextStage.getReadOnlyProperty();
    }

    void setNextStage(Stage nextStage) {
        Objects.requireNonNull(nextStage);
        this.nextStage.set(nextStage);
    }

    public int getMinutesBeforeTarget() {
        return minutesBeforeTarget.get();
    }

    public ReadOnlyIntegerProperty minutesBeforeTargetProperty() {
        return minutesBeforeTarget.getReadOnlyProperty();
    }

    void setMinutesBeforeTarget(int minutesBeforeTarget) {
        this.minutesBeforeTarget.set(minutesBeforeTarget);
    }

    public boolean isRunning() {
        return running.get();
    }

    public ReadOnlyBooleanProperty runningProperty() {
        return running.getReadOnlyProperty();
    }

    void setRunning(boolean running) {
        this.running.set(running);
    }
}

File: src/main/kotlin/io/github/dylmeadows/eontimer/model/TimerState.kt
package io.github.dylmeadows.eontimer.model

import io.github.dylmeadows.eontimer.util.getValue
import io.github.dylmeadows.eontimer.util.setValue
import javafx.beans.property.BooleanProperty
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class TimerState {
    val totalTimeProperty: ObjectProperty<Duration> = SimpleObjectProperty(Duration.ZERO)
    var totalTime: Duration by totalTimeProperty

    val totalElapsedProperty: ObjectProperty<Duration> = SimpleObjectProperty(Duration.ZERO)
    var totalElapsed: Duration by totalElapsedProperty

    val currentStageProperty: ObjectProperty<Duration> = SimpleObjectProperty(Duration.ZERO)
    var currentStage: Duration by currentStageProperty

    val nextStageProperty: ObjectProperty<Duration> = SimpleObjectProperty(Duration.ZERO)
    var nextStage: Duration by nextStageProperty

    val currentRemainingProperty: ObjectProperty<Duration> = SimpleObjectProperty(Duration.ZERO)
    var currentRemaining: Duration by currentRemainingProperty

    val runningProperty: BooleanProperty = SimpleBooleanProperty(false)
    var running: Boolean by runningProperty
}
